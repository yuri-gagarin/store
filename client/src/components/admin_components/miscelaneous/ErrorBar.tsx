import React from "react";
import { Dimmer, Loader, Segment } from "semantic-ui-react";

const ErrorBar: React.FC<{}> = (props): JSX.Element => {

  return (
    <div style={{ height: "50px", width: "100%" }}>
      <Segment>
        <Dimmer active>
          <Loader>Error Occured</Loader>
        </Dimmer>
      </Segment>
    </div>
  );
};

export default ErrorBar;
