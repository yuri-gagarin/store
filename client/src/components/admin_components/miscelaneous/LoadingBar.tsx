import React from "react";
import { Dimmer, Loader, Segment } from "semantic-ui-react";

interface Props {
  message?: string;
}
const LoadingBar: React.FC<Props> = ({ message }): JSX.Element => {

  return (
    <div style={{ height: "50px", width: "100%" }}>
      <Segment>
        <Dimmer active>
          <Loader>{message ? message : "Loading"}</Loader>
        </Dimmer>
      </Segment>
    </div>
  );
};

export default LoadingBar;
