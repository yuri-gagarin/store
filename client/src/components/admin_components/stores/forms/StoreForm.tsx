import React, { useState, useEffect, useRef }  from "react";
import { Button, Form, TextArea } from "semantic-ui-react";
// type definitions //
import { StoreFormState } from "../type_definitions/storeTypes";

type Props = {
  title: string;
  description: string;
  newForm: boolean;
  handleCreateStore(title: string, description: string): void;
  handleUpdateStore(title: string, description: string): void;
}

const StoreForm: React.FC<Props> = ({ title, description, newForm, handleCreateStore, handleUpdateStore }): JSX.Element => {
  // local state //
  const [ formState, setFormState ] = useState<StoreFormState>({ title, description });
  const storeFormRef = useRef<HTMLDivElement>(document.createElement("div"));
  // local event listeners - handlers //
  const handleTitleChange = (e: React.ChangeEvent<HTMLInputElement>):void => {
    setFormState({
      ...formState,
      title: e.target.value
    });
  };
  const hadnleDescriptionChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setFormState({
      ...formState,
      description: e.target.value
    });
  };
  const handleSubmit = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (newForm) {
      handleCreateStore(formState.title, formState.description);
    } else {
      handleUpdateStore(formState.title, formState.description);
    }
  };
  // lifecycle hooks //
  useEffect(() => {
    if (storeFormRef.current) {
      const elem = storeFormRef.current.getBoundingClientRect();
      window.scrollTo({
        top: elem.bottom,
        behavior: "smooth"
      })
    }
  }, [storeFormRef]);
  
  return (
    <div id="createStoreFormHolder" ref={storeFormRef}>
      <Form id="createStoreForm">
        <Form.Field>
          <label>Store title</label>
          <input 
            id="adminStoreFormTitleInput"
            onChange={handleTitleChange} 
            placeholder="Store title here ..." 
            value={formState.title}
          />
        </Form.Field>
        <Form.Field
          id="adminStoreFormDescInput"
          control={TextArea}
          label='Store Description'
          onChange={hadnleDescriptionChange}
          placeholder='Store description here...'
          value={formState.description}
         />
         {
           newForm 
            ? <Button type='submit' id="adminStoreFormCreateBtn" onClick={handleSubmit} content= "Create Store" />
            : <Button type='submit' id="adminStoreFormUpdateBtn" onClick={handleSubmit} content= "Update Store" />
         }  
      </Form>
    </div>
  );
};

export default StoreForm;