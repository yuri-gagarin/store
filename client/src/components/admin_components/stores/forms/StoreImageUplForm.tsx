import React, { useState, useContext } from "react";
import { Button, Form } from "semantic-ui-react";
// css imports //
import "./css/storeImgUploadForm.css";
// actions //
import { uploadStoreImage } from "../actions/APIstoreActions";
import { Store } from "../../../../state/Store";

const StoreImageUplForm: React.FC<{}> = (props): JSX.Element => {
  const { state, dispatch } = useContext(Store);
  const { loading, currentStoreData } = state.storeState;
  // local state //
  const [ file, setFile ] = useState<File | null>(null);

  // event handlers and listeners //
  const handleButtonClick = () => {

  };
  const uploadFile = () => {
    if (file) {
      const formData = new FormData();
      const { _id } = currentStoreData;
      formData.append("storeImage", file);
      uploadStoreImage(_id, formData, state, dispatch)
        .then((_) => {
          setFile(null);
        })
        .catch((_) => {
          // log error? show error message ? //
        });
    }
  };
  const cancelFile = () => {
    setFile(null);
  };
  const fileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (e.target.files) {
      setFile(e.target.files[0])
    }
  };

  return (
    <div id="storeImgUplFormHolder">
      <div><p>Image Uploader</p></div>
      <Form id="storeImgUploadForm">
        {
          !file ?
          <div id="storeImgInputControlls"> 
            <Button
              id="selectStoreImgBtn"
              as="label"
              content="Choose Image"
              labelPosition="left"
              htmlFor="storeImgFile"
              icon="file"
              onClick={handleButtonClick}
            />
            <input type="file" id="storeImgFile" hidden onChange={fileChange} />
          </div>
          : null
        }
        
        {
          file ?  
          <div id="storeImgUploadControlls">
            <Button 
              id="storeImgCancelBtn"
              as="label"
              content="Cancel"
              labelPosition="left"
              icon="cancel"
              onClick={cancelFile}
            />
            <Button 
              id="storeImgUploadBtn"
              as="label"
              content="Upload"
              icon="upload"
              onClick={uploadFile}
              loading={loading}
            />
          </div>
          : null
        }
       

       <span>{file ? file.name : "No File"}</span>
      </Form>
    </div>
  );
};

export default StoreImageUplForm;

