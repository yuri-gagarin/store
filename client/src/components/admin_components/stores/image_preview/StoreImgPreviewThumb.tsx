import React from "react";
import { Icon, Image } from "semantic-ui-react";
// css  imports //
import "./css/storeImgPreviewThumb.css";


// image preview thumbnail //
interface ImgPreviewProps {
  _id: string;
  url: string;
  handleDelete(_id: string): void;
}
export const StoreImgPreviewThumb: React.FC<ImgPreviewProps> = ({ _id, url, handleDelete }): JSX.Element => {
  return (
    <div className="storeImgPreviewThumb">
      <Image size="small" src={url}>

      </Image>
      <Icon 
        name="trash alternate" 
        className="storeImgPreviewThumbDelete" 
        onClick={() => handleDelete(_id)}>
      </Icon>
    </div>
  );
};

export default StoreImgPreviewThumb;




