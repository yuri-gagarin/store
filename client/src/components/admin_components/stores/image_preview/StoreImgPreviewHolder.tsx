import React from "react";
// css //
import "./css/storeImgPreviewHolder.css"
// additional components //
import StoreImgPreviewThumb from "./StoreImgPreviewThumb";
// api actions //
import { deleteStoreImage } from "../actions/APIstoreActions";
import { IGlobalAppState, AppAction } from "../../../../state/Store";

interface Props {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>
}
const StoreImgPreviewHolder: React.FC<Props> = ({ state, dispatch }): JSX.Element => {
  const { images } = state.storeState.currentStoreData;

  // delete image handler
  const handleDelete = (_id: string) => {
    deleteStoreImage(_id, state, dispatch);
  };

  return (
    images.length > 0 ?
    <div id="storeImgPreviewHolder">
        <div className="storeImgHolderTitle">Uploaded Images</div>
        {
          images.map((image) => {
            return (
              <StoreImgPreviewThumb 
                key={image._id} 
                _id={image._id} 
                url={image.url} 
                handleDelete={handleDelete}
              />
            );
          })
        }
      </div>
    :
    <div id="storeImgPreviewNone">
      <h3>No images uploaded... you should upload some... </h3>
    </div>
  );
};

export default StoreImgPreviewHolder;