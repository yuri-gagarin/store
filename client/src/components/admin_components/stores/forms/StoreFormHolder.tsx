import React, { useState, useEffect } from "react";
import { Button, Grid } from "semantic-ui-react";
// css imports //
import "./css/storeFormHolder.css";
// additional components //
import StoreForm from "./StoreForm";
import StoreImgPreviewHolder from "../image_preview/StoreImgPreviewThumbs";
import StoreImageUplForm from "./StoreImageUplForm";
// state //
import { IGlobalAppState, AppAction } from "../../../../state/Store";
// api actions //
import { createStore, editStore } from "../actions/APIstoreActions";
// helpers //
import { ConvertDate } from "../../../helpers/displayHelpers";

interface Props {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}

const StoreFormHolder: React.FC<Props> = ({ state, dispatch }): JSX.Element => {
  const [ formOpen, setFormOpen ] = useState<boolean>(false);
  const [ imgUpload, setImgUpload ] = useState<boolean>(false);
  const [ newForm, setNewForm ] = useState<boolean>(true);

  const { currentStoreData } = state.storeState;
  const { title, description } = currentStoreData;

  const handleCreateStore = (title: string, description: string): void => {
    const storePams = {
      title,
      description,
      storeImages: currentStoreData.images
    };
    createStore(storePams, dispatch)
      .then((success) => {
        if (success) {
          // store created //
          setImgUpload(true);
        } else {
          console.error("error");
        }
      });
  };

  const handleUpdateStore = (title: string, description: string): void => {
    const storeParams = {
      title, description, storeImages: currentStoreData.images
    }
    editStore(currentStoreData._id, storeParams, dispatch, state)
      .then((success) => {
        if (success) {
          setFormOpen(false);
        }
      })
  }

  const handleFormOpen = () => {
    setFormOpen(!formOpen);
  };
  useEffect(() => {
    console.log("form loaded")
  }, []);

  useEffect(() => {
    if (!formOpen) {
      window.scrollTo({ top: 0, behavior: "smooth" });
    }
  }, [formOpen])
  useEffect(() => {
    if (title && description) {
      setNewForm(false);
    }
  }, [title, description]);


  return (
    <div id="storeFormHolder">
      {
        !newForm ?
          <React.Fragment>
            <Grid.Row>
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <h1>Details</h1>
                <div className="storeFormHolderDetails">
                  <h3>Store title:</h3>
                  <p>{currentStoreData.title}</p>
                </div>
                <div className="storeFormHolderDetails">
                  <h3>Store description:</h3>
                  <p>{currentStoreData.description}</p>
                </div>
                <div className="storeFormHolderTimestamps">
                  <span>Created At: {ConvertDate.international(currentStoreData.createdAt)}</span>
                  <span>Edited At: {ConvertDate.international(currentStoreData.editedAt)}</span>
                </div>
              </Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <StoreImgPreviewHolder state={state} dispatch={dispatch} />
              </Grid.Column>
            </Grid.Row>
            <StoreImageUplForm />
          </React.Fragment>
          : null
      }
      <Grid.Row>
        <Grid.Column mobile={16} tablet={15} computer={14}>
          <Button  id="storeFormToggleBtn" onClick={handleFormOpen} content={ !formOpen ? "Open Form" : "Close Form"}></Button>
          {
            formOpen ? <StoreForm 
                        title={title} 
                        description={description} 
                        handleCreateStore={handleCreateStore}
                        handleUpdateStore={handleUpdateStore}
                      /> 
                      : null
          }
         
        </Grid.Column>
      </Grid.Row>
    </div>
  );
}

export default StoreFormHolder;

