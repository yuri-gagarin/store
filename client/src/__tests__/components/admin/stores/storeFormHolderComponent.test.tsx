import React from "react"
import { Button, Icon } from "semantic-ui-react";
// testing utils
import { mount, ReactWrapper } from "enzyme";
import moxios from "moxios";
import { act } from "react-dom/test-utils";
// client routing //
import { MemoryRouter as Router } from "react-router-dom";
import { AdminStoreRoutes } from "../../../../routes/adminRoutes";
// component imports //
import StoreFormHolder from "../../../../components/admin_components/stores/forms/StoreFormHolder";
import StoreForm from "../../../../components/admin_components/stores/forms/StoreForm";
import StoreImageUplForm from "../../../../components/admin_components/stores/forms/StoreImageUplForm";
import StoreImgPreviewHolder from "../../../../components/admin_components/stores/image_preview/StoreImgPreviewHolder";
import StoreImgPreviewThumb from "../../../../components/admin_components/stores/image_preview/StoreImgPreviewThumb";
import LoadingBar from "../../../../components/admin_components/miscelaneous/LoadingBar";
import ErrorBar from "../../../../components/admin_components/miscelaneous/ErrorBar";
// state React.Context //
import { IGlobalAppState, TestStateProvider } from "../../../../state/Store";
// helpers //
import { createMockStores, setMockStoreState } from "../../../../test_helpers/storeHelpers";
import { generateCleanState } from "../../../../test_helpers/miscHelpers";

describe("StoreFormHolder Component tests", () => {
  let wrapper: ReactWrapper; 
  
  describe("Default Form Holder state",  () => {

    beforeAll(() => {
      window.scrollTo = jest.fn();
      wrapper = mount(
        <Router keyLength={0}>
          <StoreFormHolder />
        </Router>
      );
    });

    it("Should Properly Mount Form Holder", () => {
      expect(wrapper.find(StoreFormHolder)).toMatchSnapshot();
    });
    it("Form Should be closed by default", () => {
      const form = wrapper.find(StoreForm);
      expect(form.length).toEqual(0);
    });
    it("Should NOT render '#adminStoreFormHolderDetails' 'div'", () => {
      const formDetails = wrapper.find(StoreFormHolder).render().find("#adminStoreFormHolderDetails");
      expect(formDetails.length).toEqual(0);
    });
    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.find(StoreFormHolder).render().find("#adminStoreFormToggleBtn");
      expect(toggleButton.length).toEqual(1);
    });

  });
  // TEST Form Holder state OPEN - NO Current Store Data //
  describe("Form Holder state OPEN - NO Current Store Data",  () => {
    let wrapper: ReactWrapper;

    beforeAll(() => {
      window.scrollTo = jest.fn();
      wrapper = mount(
        <Router keyLength={0}>
          <TestStateProvider>
            <StoreFormHolder />
          </TestStateProvider>
        </Router>
      );
    });
    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.render().find('#adminStoreFormToggleBtn');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should NOT render '#adminStoreFormHolderDetails' 'div'", () => {
      const formDetails = wrapper.find(StoreFormHolder).render().find("#adminStoreFormHolderDetails");
      expect(formDetails.length).toEqual(0);
    });
    it("Should Properly Mount Form Holder, respond to '#storeToggleBtn' click", () => {
      const toggleButton = wrapper.find("#adminStoreFormToggleBtn");
      toggleButton.at(0).simulate("click")
      // open button clicked //
      //wrapper.update()
      expect(wrapper).toMatchSnapshot();
    });
  
    it("Should have a Form Create Button", () => {
      const toggleButton = wrapper.render().find('#adminStoreFormCreateBtn');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Form rendered after toggle button click", () => {
      const form = wrapper.find(StoreForm);
      expect(form.length).toEqual(1);
    });
    it("Should NOT have the Image Preview rendered", () => {
      const imgPreviewHolder = wrapper.find(StoreImgPreviewHolder);
      expect(imgPreviewHolder.length).toEqual(0);
    });
    it("Should NOT have the Image Upload Form rendered", () => {
      const imgUploadForm = wrapper.find(StoreImageUplForm);
      expect(imgUploadForm.length).toEqual(0);
    });
  });
  // END Form Holder state OPEN - NO Current Store Data //
  // TEST Form Holder state OPEN - WITH Current Store Data - NO IMAGES //
  describe("Form Holder state OPEN - WITH Current Store Data - NO IMAGES",  () => {
    let wrapper: ReactWrapper; let state: IGlobalAppState;

    beforeAll(() => {
      window.scrollTo = jest.fn();
      state = setMockStoreState({ currentStore: true });
      wrapper = mount(
        <Router keyLength={0}>
          <TestStateProvider mockState={state}>
            <StoreFormHolder />
          </TestStateProvider>
        </Router>
      );
      wrapper.update()
    });

    it("Should Properly Mount Form Holder", () => {
      expect(wrapper.find("#storeFormHolder").length).toEqual(1);
    });
    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.render().find('#adminStoreFormToggleBtn');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should  NOT initially render the 'StoreForm' component after toggle button click", () => {
      const form = wrapper.find(StoreForm);
      expect(form.length).toEqual(0);
    });
    it("Should render '#adminStoreFormHolderDetails' 'div'", () => {
      const formDetails = wrapper.find(StoreFormHolder).render().find("#adminStoreFormHolderDetails");
      expect(formDetails.length).toEqual(1);
    });
    it("Should correctly render data in '.adminStoreFormHolderDetailsItem' <div>(s)", () => {
      const detailsDivs = wrapper.find(StoreFormHolder).find('.adminStoreFormHolderDetailsItem');
      const { currentStoreData } = state.storeState;
      expect(detailsDivs.length).toEqual(2);
      expect(detailsDivs.at(0).find('p').text()).toEqual(currentStoreData.title);
      expect(detailsDivs.at(1).find('p').text()).toEqual(currentStoreData.description);
    });
    it("Should render 'StoreForm' component after '#adminStoreFormToggleBtn' click event", () => {
      wrapper.find("#adminStoreFormToggleBtn").at(0).simulate("click");
      wrapper.update();
      expect(wrapper.find(StoreForm).length).toEqual(1);
    })
    it("Should have a Form Update Button", () => {
      const toggleButton = wrapper.render().find('#adminStoreFormUpdateBtn');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Image Preview rendered", () => {
      const imgPreviewHolder = wrapper.find(StoreImgPreviewHolder);
      expect(imgPreviewHolder.length).toEqual(1);
    });
    it("Should NOT render any preview images", () => {
      const previewThumb = wrapper.find(StoreImgPreviewThumb);
      expect(previewThumb.length).toEqual(0);
    })
    it("Should have the Image Upload Form rendered", () => {
      const imgUploadForm = wrapper.find(StoreImageUplForm);
      expect(imgUploadForm.length).toEqual(1);
    });
  });
  // END Form Holder state OPEN - WITH Current Store Data - NO IMAGES //
  // TEST Form Holder state OPEN - WITH Current Store Data - WITH IMAGES //
  describe("Form Holder state OPEN - WITH Current Store Data - WITH IMAGES",  () => {
    let state: IGlobalAppState; let wrapper: ReactWrapper;

    beforeAll(() => {
      window.scrollTo = jest.fn();
      state = setMockStoreState({ currentStore: true, storeImages: 3 });
      wrapper = mount(
        <Router keyLength={0}>
          <TestStateProvider mockState={state}>
            <StoreFormHolder />
          </TestStateProvider>
        </Router>
      );
    });

    it("Should Properly Mount Form Holder", () => {
      expect(wrapper.find("#storeFormHolder").length).toEqual(1);
    });
    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.render().find('#adminStoreFormToggleBtn');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should NOT have the 'StoreForm' component initially rendered", () => {
      expect(wrapper.find(StoreForm).length).toEqual(0)
    });
    it("Should render '#adminStoreFormHolderDetails' 'div'", () => {
      const formDetails = wrapper.find(StoreFormHolder).render().find("#adminStoreFormHolderDetails");
      expect(formDetails.length).toEqual(1);
    })
    it("Should correctly render data in '.adminStoreFormHolderDetailsItem' <div>(s)", () => {
      const detailsDivs = wrapper.find(StoreFormHolder).find('.adminStoreFormHolderDetailsItem');
      const { currentStoreData } = state.storeState;
      expect(detailsDivs.length).toEqual(2);
      expect(detailsDivs.at(0).find('p').text()).toEqual(currentStoreData.title);
      expect(detailsDivs.at(1).find('p').text()).toEqual(currentStoreData.description);
    });
    it("Should render 'StoreForm' componet after '#adminStoreFormToggleBtn' click event", () => {
      const toggleBtn = wrapper.find(StoreFormHolder).find("#adminStoreFormToggleBtn");
      toggleBtn.at(0).simulate("click");
      // assert correct rendering //
      const form = wrapper.find(StoreForm);
      expect(form.length).toEqual(1);
    });
    it("Should have a Form Update Button", () => {
      const toggleButton = wrapper.render().find('#adminStoreFormUpdateBtn');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Image Preview rendered", () => {
      const imgPreviewHolder = wrapper.find(StoreImgPreviewHolder);
      expect(imgPreviewHolder.length).toEqual(1);
    });
    it("Should render a correct number of preview images", () => {
      const previewThumb = wrapper.find(StoreImgPreviewThumb);
      const numberOfImages = state.storeState.currentStoreData.images.length;
      expect(previewThumb.length).toEqual(numberOfImages);
    })
    it("Should have the Image Upload Form rendered", () => {
      const imgUploadForm = wrapper.find(StoreImageUplForm);
      expect(imgUploadForm.length).toEqual(1);
    });
  });
  // END Form Holder state OPEN - WITH Current Store Data - WITH IMAGES //
  // TEST Form Holder state OPEN - MOCK Submit action //
  describe("Form Holder state OPEN - MOCK Submit action",  () => {
    let state: IGlobalAppState; let wrapper: ReactWrapper;

    beforeAll( async () => {
      window.scrollTo = jest.fn();
      state = generateCleanState();
      // mount and wait //
      wrapper = mount(
        <Router initialEntries={[AdminStoreRoutes.EDIT_ROUTE]} keyLength={0} >
          <TestStateProvider mockState={state}>
            <StoreFormHolder />
          </TestStateProvider>
        </Router>
      );
    });
    afterAll(() => {
      moxios.uninstall();
    });
    
    it("Should have a submit button", () => {
      wrapper.find("#adminStoreFormToggleBtn").at(0).simulate("click");
      wrapper.update();
      const adminStoreFormCreate = wrapper.find("#adminStoreFormCreateBtn").at(0);
      expect(adminStoreFormCreate.length).toEqual(1)
    });
    it("Should handle the 'handleCreateStoreAction, show 'LoadingBar' Component", async () => {
      const promise = Promise.resolve();
      moxios.install();
      moxios.stubRequest("/api/stores/create", {
        status: 200,
        response: {
          responseMsg: "All Good",
          newStore: createMockStores(1)[0]
        }
      });
      const adminStoreFormCreate = wrapper.find("#adminStoreFormCreateBtn").at(0);
      // console.log(wrapper.find(StoreFormHolder).debug());
      adminStoreFormCreate.simulate("click");
      await act( async () => promise);
      expect(wrapper.find(LoadingBar).length).toEqual(1);
    });
    it("Should NOT show the 'LoadingBar' Component after successful API call", () => {
      wrapper.update();
      expect(wrapper.find(LoadingBar).length).toEqual(0);
    });
    it("Should NOT show the 'StoreForm' Component after successful API call", () => {
      expect(wrapper.find(StoreForm).length).toEqual(0);
    });
  });
  // END Form Holder state OPEN - MOCK Submit action //
  // TEST StoreFormHolder mock submit action with an API error returned //
  describe("Form Holder state OPEN - MOCK Submit action ERROR returned",  () => {
    let state: IGlobalAppState; let wrapper: ReactWrapper;
    let mockStore: IStoreData;
    const error = new Error("Am error occured");

    beforeAll( async () => {
      window.scrollTo = jest.fn();
      state = generateCleanState();
      const currentStore = mockStore = createMockStores(1)[0];
      state.storeState.currentStoreData = currentStore;
      // mount and wait //
      wrapper = mount(
        <Router initialEntries={[AdminStoreRoutes.EDIT_ROUTE]} keyLength={0} >
          <TestStateProvider mockState={state}>
            <StoreFormHolder />
          </TestStateProvider>
        </Router>
      );
      const toggleBtn = wrapper.find(StoreFormHolder).find("#adminStoreFormToggleBtn");
      toggleBtn.at(0).simulate("click");
    });
    afterAll(() => {
      moxios.uninstall();
    });
    
    it("Should handle the 'handleCreateStoreAction, show 'LoadingBar' Component", async () => {
      const promise = Promise.resolve();
      const url = `/api/stores/update/${mockStore._id}`
      moxios.install();
      moxios.stubRequest(url, {
        status: 500,
        response: {
          responseMsg: "Error",
          error: error
        }
      });
      const adminStoreFormCreate = wrapper.find("#adminStoreFormUpdateBtn").at(0);
      adminStoreFormCreate.simulate("click");
      await act( async () => promise);
      expect(wrapper.find(LoadingBar).length).toEqual(1);
      expect(moxios.requests.mostRecent().url).toEqual(url);
    });
    it("Should NOT render the 'LoadingBar' Component after an error in API call", () => {
      wrapper.update();
      expect(wrapper.find(LoadingBar).length).toEqual(0);
    });
    it("Should render the 'ErrorBar' Component after an error in API call", () => {
      expect(wrapper.find(ErrorBar).length).toEqual(1);
    });
    it("Should show the 'StoreForm' Component after an error in API call", () => {
      expect(wrapper.find(StoreForm).length).toEqual(1);
    });
    it("Should properly dissmiss the 'ErrorBar' component with button click", () => {
      const dismissErrorIcon = wrapper.find(StoreFormHolder).find(ErrorBar).find(Icon);
      // simulate the dismissErrorIcon click //
      dismissErrorIcon.simulate("click");
      expect(wrapper.find(StoreFormHolder).find(ErrorBar).length).toEqual(0);
      expect(wrapper.find(StoreFormHolder).find(LoadingBar).length).toEqual(0);
    })
  });

  
});