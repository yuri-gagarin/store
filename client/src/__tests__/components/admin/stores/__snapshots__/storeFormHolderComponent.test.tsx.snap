// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreFormHolder Component tests Default Form Holder state Should Properly Mount Form Holder 1`] = `
<withRouter(StoreFormHolder)>
  <StoreFormHolder
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    location={
      Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      }
    }
    match={
      Object {
        "isExact": true,
        "params": Object {},
        "path": "/",
        "url": "/",
      }
    }
  >
    <div
      id="storeFormHolder"
    >
      <GridRow>
        <div
          className="row"
        >
          <GridColumn
            computer={14}
            mobile={16}
            tablet={15}
          >
            <div
              className="fourteen wide computer sixteen wide mobile fifteen wide tablet column"
            >
              <Button
                as="button"
                content="Open Form"
                id="adminStoreFormToggleBtn"
                onClick={[Function]}
              >
                <Ref
                  innerRef={
                    Object {
                      "current": <button
                        class="ui button"
                        id="adminStoreFormToggleBtn"
                      >
                        Open Form
                      </button>,
                    }
                  }
                >
                  <RefFindNode
                    innerRef={
                      Object {
                        "current": <button
                          class="ui button"
                          id="adminStoreFormToggleBtn"
                        >
                          Open Form
                        </button>,
                      }
                    }
                  >
                    <button
                      className="ui button"
                      id="adminStoreFormToggleBtn"
                      onClick={[Function]}
                    >
                      Open Form
                    </button>
                  </RefFindNode>
                </Ref>
              </Button>
            </div>
          </GridColumn>
        </div>
      </GridRow>
    </div>
  </StoreFormHolder>
</withRouter(StoreFormHolder)>
`;

exports[`StoreFormHolder Component tests Form Holder state OPEN - NO Current Store Data Should Properly Mount Form Holder, respond to '#storeToggleBtn' click 1`] = `
<MemoryRouter
  keyLength={0}
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Component>
      <withRouter(StoreFormHolder)>
        <StoreFormHolder
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "canGo": [Function],
              "createHref": [Function],
              "entries": Array [
                Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
              ],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "index": 0,
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
        >
          <div
            id="storeFormHolder"
          >
            <GridRow>
              <div
                className="row"
              >
                <GridColumn
                  computer={14}
                  mobile={16}
                  tablet={15}
                >
                  <div
                    className="fourteen wide computer sixteen wide mobile fifteen wide tablet column"
                  >
                    <Button
                      as="button"
                      content="Close Form"
                      id="adminStoreFormToggleBtn"
                      onClick={[Function]}
                    >
                      <Ref
                        innerRef={
                          Object {
                            "current": <button
                              class="ui button"
                              id="adminStoreFormToggleBtn"
                            >
                              Close Form
                            </button>,
                          }
                        }
                      >
                        <RefFindNode
                          innerRef={
                            Object {
                              "current": <button
                                class="ui button"
                                id="adminStoreFormToggleBtn"
                              >
                                Close Form
                              </button>,
                            }
                          }
                        >
                          <button
                            className="ui button"
                            id="adminStoreFormToggleBtn"
                            onClick={[Function]}
                          >
                            Close Form
                          </button>
                        </RefFindNode>
                      </Ref>
                    </Button>
                    <StoreForm
                      description=""
                      handleCreateStore={[Function]}
                      handleUpdateStore={[Function]}
                      newForm={true}
                      title=""
                    >
                      <div
                        id="createStoreFormHolder"
                      >
                        <Form
                          as="form"
                          id="createStoreForm"
                        >
                          <form
                            className="ui form"
                            id="createStoreForm"
                            onSubmit={[Function]}
                          >
                            <FormField>
                              <div
                                className="field"
                              >
                                <label>
                                  Store title
                                </label>
                                <input
                                  id="adminStoreFormTitleInput"
                                  onChange={[Function]}
                                  placeholder="Store title here ..."
                                  value=""
                                />
                              </div>
                            </FormField>
                            <FormField
                              control={[Function]}
                              id="adminStoreFormDescInput"
                              label="Store Description"
                              onChange={[Function]}
                              placeholder="Store description here..."
                              value=""
                            >
                              <div
                                className="field"
                              >
                                <label
                                  htmlFor="adminStoreFormDescInput"
                                >
                                  Store Description
                                </label>
                                <TextArea
                                  aria-describedby={null}
                                  as="textarea"
                                  id="adminStoreFormDescInput"
                                  onChange={[Function]}
                                  placeholder="Store description here..."
                                  rows={3}
                                  value=""
                                >
                                  <Ref
                                    innerRef={
                                      Object {
                                        "current": <textarea
                                          id="adminStoreFormDescInput"
                                          placeholder="Store description here..."
                                          rows="3"
                                        />,
                                      }
                                    }
                                  >
                                    <RefFindNode
                                      innerRef={
                                        Object {
                                          "current": <textarea
                                            id="adminStoreFormDescInput"
                                            placeholder="Store description here..."
                                            rows="3"
                                          />,
                                        }
                                      }
                                    >
                                      <textarea
                                        aria-describedby={null}
                                        id="adminStoreFormDescInput"
                                        onChange={[Function]}
                                        onInput={[Function]}
                                        placeholder="Store description here..."
                                        rows={3}
                                        value=""
                                      />
                                    </RefFindNode>
                                  </Ref>
                                </TextArea>
                              </div>
                            </FormField>
                            <Button
                              as="button"
                              content="Create Store"
                              id="adminStoreFormCreateBtn"
                              onClick={[Function]}
                              type="submit"
                            >
                              <Ref
                                innerRef={
                                  Object {
                                    "current": <button
                                      class="ui button"
                                      id="adminStoreFormCreateBtn"
                                      type="submit"
                                    >
                                      Create Store
                                    </button>,
                                  }
                                }
                              >
                                <RefFindNode
                                  innerRef={
                                    Object {
                                      "current": <button
                                        class="ui button"
                                        id="adminStoreFormCreateBtn"
                                        type="submit"
                                      >
                                        Create Store
                                      </button>,
                                    }
                                  }
                                >
                                  <button
                                    className="ui button"
                                    id="adminStoreFormCreateBtn"
                                    onClick={[Function]}
                                    type="submit"
                                  >
                                    Create Store
                                  </button>
                                </RefFindNode>
                              </Ref>
                            </Button>
                          </form>
                        </Form>
                      </div>
                    </StoreForm>
                  </div>
                </GridColumn>
              </div>
            </GridRow>
          </div>
        </StoreFormHolder>
      </withRouter(StoreFormHolder)>
    </Component>
  </Router>
</MemoryRouter>
`;
